In [125]:
import pandas as pd

from itertools import combinations

# Data preparation for my analisys

## NFL injuries

In [126]:

nfl = pd.read_csv("../datasets/raw/nfl_injuries_2009_2021.csv")
nba = pd.read_csv("../datasets/raw/nba_injuries_2010-2020.csv")
nhl = pd.read_csv("../datasets/raw/nhl_injuries.csv")
mlb = pd.read_csv("../datasets/raw/mlb_injury.csv")

In [127]:
nfl.head()

,Unnamed: 0,Player_id,Player,year,team,games_in_season,num_games_missing,num_games_injured,injury_types,earliest_injury,latest_injury
0,all_info,AdamMi98,"Adams,Michael",2009,crd,16,0,3,hamstring,8.0,14.0
1,all_info.1,BankJa00,"Banks,Jason",2009,crd,16,0,0,NaN,NaN,NaN
2,all_info.2,BerrBe99,"Berry,Bertrand",2009,crd,16,0,1,illness,13.0,13.0
3,all_info.3,BoldAn00,"Boldin,Anquan",2009,crd,16,1,6,hamstring ankle,1.0,9.0
4,all_info.4,BranAl99,"Branch,Alan",2009,crd,16,0,1,ankle,13.0,13.0


In [128]:
nfl.info()

<class 'pandas.DataFrame'>
RangeIndex: 17389 entries, 0 to 17388
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         17389 non-null  str    
 1   Player_id          17389 non-null  str    
 2   Player             17389 non-null  str    
 3   year               17389 non-null  int64  
 4   team               17389 non-null  str    
 5   games_in_season    17389 non-null  int64  
 6   num_games_missing  17389 non-null  int64  
 7   num_games_injured  17389 non-null  int64  
 8   injury_types       15910 non-null  str    
 9   earliest_injury    17087 non-null  float64
 10  latest_injury      17087 non-null  float64
dtypes: float64(2), int64(4), str(5)
memory usage: 1.5 MB


In [129]:
nfl.describe()

,year,games_in_season,num_games_missing,num_games_injured,earliest_injury,latest_injury
count,17389.000000,17389.000000,17389.000000,17389.000000,17087.000000,17087.000000
mean,2014.747829,15.723906,4.110817,5.640002,5.813308,12.327969
std,3.644730,1.234834,5.132237,4.898407,4.548916,4.439140
min,2009.000000,9.000000,0.000000,0.000000,1.000000,1.000000
25%,2012.000000,16.000000,0.000000,2.000000,2.000000,9.000000
50%,2015.000000,16.000000,2.000000,4.000000,5.000000,14.000000
75%,2018.000000,16.000000,6.000000,8.000000,9.000000,16.000000
max,2021.000000,16.000000,16.000000,16.000000,16.000000,16.000000


In [130]:
nfl.isnull().sum()

Unnamed: 0              0
Player_id               0
Player                  0
year                    0
team                    0
games_in_season         0
num_games_missing       0
num_games_injured       0
injury_types         1479
earliest_injury       302
latest_injury         302
dtype: int64

In [131]:
nfl.shape

(17389, 11)

In [132]:
print("NFL Dataset Duplicates:", nfl.duplicated().sum())
print("\nUnique values per column:\n", nfl.nunique())

NFL Dataset Duplicates: 0

Unique values per column:
 Unnamed: 0           17389
Player_id             6023
Player                5947
year                    13
team                    32
games_in_season          4
num_games_missing       17
num_games_injured       17
injury_types          3760
earliest_injury         16
latest_injury           16
dtype: int64


In [133]:
nfl.columns.value_counts()

Unnamed: 0           1
Player_id            1
Player               1
year                 1
team                 1
games_in_season      1
num_games_missing    1
num_games_injured    1
injury_types         1
earliest_injury      1
latest_injury        1
Name: count, dtype: int64

**NFL Dataset Observations:**
*   **Description:** This dataset tracks injuries in the National Football League (NFL) over a span of multiple seasons (2009-2021). It includes information on the players, specific games, the type of injury sustained, and the resulting missed playing time.
*   **Data Source:** https://github.com/clementzach/BTS_260_project/blob/master/Data/data_build/injuries_2009_2021.csv
*   **Limitations:** The dataset does not detail multiple separate injury events sequentially occurring within the same match. Certain vague injury categorizations may require domain knowledge to decipher.
*   **Columns:** The dataset consists of 11 columns, including key features such as `injury_types`, `num_games_missing`, `num_games_injured`, and `year`.


**Additional Insights:**
*   **Duplicates:** 0 duplicate rows found.
*   **Unique Values:** High cardinality in `injury_types` (3760 unique values) indicates we might need to group or normalize the strings (e.g., lowercase, stemming) to avoid feature explosion.
* **Injury type column?** `injury_types`
* **Games missed/recovery/duration?** `num_games_missing`, `num_games_injured`
* **How are injuries described?** Described categorically in `injury_types`.
* **Season/year?** `year`
* **Cleanest/most chaotic?** Fairly clean and structured. Size: 17,389 rows, 11 columns.

## NBA injuries

In [134]:
nba.head()

,Date,Team,Acquired,Relinquished,Notes
0,2010-10-03,Bulls,NaN,Carlos Boozer,fractured bone in right pinky finger (out inde...
1,2010-10-06,Pistons,NaN,Jonas Jerebko,torn right Achilles tendon (out indefinitely)
2,2010-10-06,Pistons,NaN,Terrico White,broken fifth metatarsal in right foot (out ind...
3,2010-10-08,Blazers,NaN,Jeff Ayres,torn ACL in right knee (out indefinitely)
4,2010-10-08,Nets,NaN,Troy Murphy,strained lower back (out indefinitely)


In [135]:
nba.info()

<class 'pandas.DataFrame'>
RangeIndex: 27105 entries, 0 to 27104
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Date          27105 non-null  str  
 1   Team          27103 non-null  str  
 2   Acquired      9542 non-null   str  
 3   Relinquished  17560 non-null  str  
 4   Notes         27105 non-null  str  
dtypes: str(5)
memory usage: 1.0 MB


In [136]:
nba.describe()

,Date,Team,Acquired,Relinquished,Notes
count,27105,27103,9542,17560,27105
unique,2103,32,1111,1156,3114
top,2014-04-16,Spurs,Dwyane Wade,Kevin Love,activated from IL
freq,76,1163,54,101,7342


In [137]:
nba.isnull().sum()

Date                0
Team                2
Acquired        17563
Relinquished     9545
Notes               0
dtype: int64

In [138]:
nba.shape

(27105, 5)

In [139]:
print("NBA Dataset Duplicates:", nba.duplicated().sum())
print("\nUnique values per column:\n", nba.nunique())

NBA Dataset Duplicates: 2

Unique values per column:
 Date            2103
Team              32
Acquired        1111
Relinquished    1156
Notes           3114
dtype: int64


In [140]:
nba.columns.value_counts()

Date            1
Team            1
Acquired        1
Relinquished    1
Notes           1
Name: count, dtype: int64

**NBA Dataset Observations:**
*   **Description:** The NBA dataset (2010-2020) tracks injuries and missed games across basketball seasons. Unlike other sports datasets, it usually aggregates injury updates in a free-text "Notes" field rather than standardized anatomical keywords, often gathered through daily injury reports.
*   **Data Source:** (https://www.kaggle.com/datasets/ghopkins/nba-injuries-2010-2018)
*   **Limitations:** Heavily relies on unstandardized free-text logs for injuries and return times. Specific injury body parts are often mixed with generalized statuses like "rest" or "illness".
*   **Columns:** The dataset is very minimal with only 5 columns, relying heavily on `Date` and `Notes`, along with `Team`, `Acquired`, and `Relinquished` to detail team movements related to injuries.



**Additional Insights:**
*   **Duplicates:** 2 duplicate rows found. These should be removed during cleaning.
*   **Unique Values:** `Notes` holds 3,114 unique text entries. Since this is the primary source of injury detail, NLP processing will be necessary to extract meaningful, categorical tags.
* **Injury type column?** No explicit categorical column; described in `Notes`.
* **Games missed/recovery/duration?** Missed games/recovery are written as free-text within the `Notes`.
* **How are injuries described?** Free-text in the `Notes`.
* **Season/year?** Mostly relies on the `Date` column.
* **Cleanest/most chaotic?** Most chaotic/unstructured due to its reliance on free-text note parsing. Size: 27,105 rows, 5 columns.

## NHL Injuries

In [141]:
nhl.head()

,Season,Team,Position,Player,Injury Type,Cap Hit,Chip,Games Missed,Games Missed.1
0,2000/01,Anaheim,F,"Kariya, Paul",Foot,NaN,NaN,16,16
1,2000/01,Anaheim,F,"Leclerc, Mike",Abdominal,NaN,NaN,13,13
2,2000/01,Anaheim,F,"Leclerc, Mike",Knee,NaN,NaN,15,15
3,2000/01,Anaheim,F,"McDonald, Andy",Concussion,NaN,NaN,7,7
4,2000/01,Anaheim,F,"McInnis, Marty",Groin,NaN,NaN,3,3


In [142]:
nhl.info()

<class 'pandas.DataFrame'>
RangeIndex: 24498 entries, 0 to 24497
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Season          24498 non-null  str    
 1   Team            24498 non-null  str    
 2   Position        24498 non-null  str    
 3   Player          24498 non-null  str    
 4   Injury Type     24498 non-null  str    
 5   Cap Hit         17666 non-null  float64
 6   Chip            17666 non-null  float64
 7   Games Missed    24498 non-null  int64  
 8   Games Missed.1  24498 non-null  int64  
dtypes: float64(2), int64(2), str(5)
memory usage: 1.7 MB


In [143]:
nhl.describe()

,Cap Hit,Chip,Games Missed,Games Missed.1
count,17666.000000,17666.000000,24498.000000,24498.000000
mean,2.966112,307.173656,8.145400,8.183566
std,2.239837,671.117390,12.125458,12.122303
min,0.300000,5.182927,1.000000,1.000000
25%,0.991667,43.602634,2.000000,2.000000
50%,2.491667,104.551390,4.000000,4.000000
75%,4.400000,285.365854,9.000000,9.000000
max,14.000000,10500.000000,82.000000,82.000000


In [144]:
nhl.isnull().sum()

Season               0
Team                 0
Position             0
Player               0
Injury Type          0
Cap Hit           6832
Chip              6832
Games Missed         0
Games Missed.1       0
dtype: int64

In [145]:
nhl.shape

(24498, 9)

In [146]:
print("NHL Dataset Duplicates:", nhl.duplicated().sum())
print("\nUnique values per column:\n", nhl.nunique())

NHL Dataset Duplicates: 0

Unique values per column:
 Season              49
Team                35
Position             9
Player            3043
Injury Type         97
Cap Hit            942
Chip              3789
Games Missed        82
Games Missed.1      82
dtype: int64


In [147]:
nhl.columns.value_counts()

Season            1
Team              1
Position          1
Player            1
Injury Type       1
Cap Hit           1
Chip              1
Games Missed      1
Games Missed.1    1
Name: count, dtype: int64

**NHL Dataset Observations:**
*   **Description:** The NHL injuries dataset tracks hockey players' time injured or put on the injured reserve. It typically organizes injuries by body parts and records the number of regular season or playoff games missed due to those specific health instances across seasons.
*   **Data Source:** (https://public.tableau.com.com/app/profile/lw3h/viz/NHLinjurydatabase/NHLInjuryDatabase)
*   **Limitations:** Includes potentially duplicate tracks for missed games making aggregation somewhat uncertain. Lacks highly granular descriptions mapping to exact anatomic locations or severity levels.
*   **Columns:** The dataset contains 9 columns, capturing identifying details like `Season`, `Team`, and `Player`, the descriptive `Injury Type`, and duration counts such as `Games Missed` and the identical `Games Missed.1`.

**Additional Insights:**
*   **Duplicates:** 0 duplicate rows found.
*   **Unique Values:** `Injury Type` has 97 unique categories, which is relatively small and manageable without heavy text parsing. The `Games Missed` and `Games Missed.1` columns seem identical, with 82 unique values each.
* **Injury type column?** `Injury Type`
* **Games missed/recovery/duration?** `Games Missed` and `Games Missed.1`
* **How are injuries described?** Categorical or short text in `Injury Type`.
* **Season/year?** Uses the `Season` column (e.g., "2000/01").
* **Cleanest/most chaotic?** Somewhat clean, but contains potentially duplicate or confusing columns (e.g., `Games Missed.1`). Size: 24,498 rows, 9 columns.

## MLB Injuries

In [148]:
mlb.head()

,date,effectiveDate,notes,player_id,team_id,year,team,team_name,name,dtd,...,cum_season_days,prev_season_days,next_game_season_days,dt,in_season_duration,date_duration,duration,t,position,__index_level_0__
0,2012-04-07 00:00:00.000000,\N,\N,110029,\N,2012,LAA,\N,"Abreu, Bobby",True,...,11,11,11,0,0,0,0,0,batter,0
1,2014-09-28 00:00:00.000000,\N,\N,110029,\N,2014,NYM,\N,"Abreu, Bobby",True,...,565,10,565,264,0,0,0,264,batter,1
2,2012-03-29 00:00:00.000000,\N,\N,112526,\N,2012,OAK,\N,Bartolo Colon,True,...,2,2,2,0,0,0,0,0,pitcher,2
3,2012-06-23 00:00:00.000000,2012-06-18,oakland athletics placed rhp bartolo colon on ...,112526,133,2012,OAK,Athletics,Bartolo Colon,False,...,88,1,98,87,10,10,15,87,pitcher,3
4,2013-08-17 00:00:00.000000,2013-08-14,oakland athletics placed rhp bartolo colon on ...,112526,133,2013,OAK,Athletics,Bartolo Colon,False,...,330,97,342,180,12,12,15,267,pitcher,4


In [149]:
mlb.describe()

,player_id,year,game,injury_span_id_orig,non_mlb_days,il_days_sum,prev_non_mlb_days,cum_season_days,prev_season_days,dt,duration,t,__index_level_0__
count,9626.000000,9626.000000,9626.000000,9626.000000,9626.000000,9626.000000,9626.000000,9626.000000,9626.000000,9626.000000,9626.000000,9626.000000,9626.000000
mean,526727.848639,2017.603678,0.234469,492.297216,15.566071,7.319655,17.653542,1100.109287,976.340017,111.706836,19.886246,582.675774,4812.500000
std,90307.879810,3.093639,0.423689,462.649540,71.727435,14.013965,68.364310,550.884836,572.444094,160.578585,43.903108,429.888121,2778.931179
min,110029.000000,2012.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,457915.000000,2015.000000,0.000000,146.000000,0.000000,0.000000,0.000000,717.250000,572.000000,11.000000,1.000000,223.000000,2406.250000
50%,537953.000000,2018.000000,0.000000,318.000000,0.000000,0.000000,0.000000,1148.000000,1025.000000,51.000000,4.000000,551.000000,4812.500000
75%,595751.000000,2021.000000,0.000000,697.000000,0.000000,10.000000,0.000000,1563.000000,1483.000000,142.000000,20.000000,886.000000,7218.750000
max,683734.000000,2022.000000,1.000000,1500.000000,1243.000000,260.000000,1243.000000,1905.000000,1904.000000,1476.000000,974.000000,1853.000000,9625.000000


In [150]:
mlb.info()

<class 'pandas.DataFrame'>
RangeIndex: 9626 entries, 0 to 9625
Data columns (total 48 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   date                   9626 non-null   str  
 1   effectiveDate          9626 non-null   str  
 2   notes                  9626 non-null   str  
 3   player_id              9626 non-null   int64
 4   team_id                9626 non-null   str  
 5   year                   9626 non-null   int64
 6   team                   9626 non-null   str  
 7   team_name              9626 non-null   str  
 8   name                   9626 non-null   str  
 9   dtd                    9626 non-null   bool 
 10  activated              9626 non-null   str  
 11  il_days                9626 non-null   str  
 12  transfer               9626 non-null   str  
 13  injury_notes           9232 non-null   str  
 14  injury_type_notes      9232 non-null   str  
 15  injury_type            9626 non-null   str  
 16 

In [151]:
mlb.isnull().sum()

date                       0
effectiveDate              0
notes                      0
player_id                  0
team_id                    0
year                       0
team                       0
team_name                  0
name                       0
dtd                        0
activated                  0
il_days                    0
transfer                   0
injury_notes             394
injury_type_notes        394
injury_type                0
injury_location_notes    394
injury_location            0
game                       0
injury_id                  0
injury_date                0
game_date                  0
game_type                  0
prev_is_game               0
next_is_game               0
prev_game_date             0
next_game_date             0
player_first_row           0
player_last_row            0
injury_span_id             0
injury_span_id_orig        0
time_since_last_game       0
non_mlb_days               0
il_days_max                0
il_days_sum   

In [152]:
mlb.shape

(9626, 48)

In [153]:
print("MLB Dataset Duplicates:", mlb.duplicated().sum())
print("\nUnique values per column:\n", mlb.nunique().head(10), "\n... (omitted remainder)")

MLB Dataset Duplicates: 0

Unique values per column:
 date             1846
effectiveDate    1501
notes            3978
player_id        1267
team_id            31
year               11
team               30
team_name          33
name             1265
dtd                 2
dtype: int64 
... (omitted remainder)


In [154]:
mlb.columns.value_counts()

date                     1
effectiveDate            1
notes                    1
player_id                1
team_id                  1
year                     1
team                     1
team_name                1
name                     1
dtd                      1
activated                1
il_days                  1
transfer                 1
injury_notes             1
injury_type_notes        1
injury_type              1
injury_location_notes    1
injury_location          1
game                     1
injury_id                1
injury_date              1
game_date                1
game_type                1
prev_is_game             1
next_is_game             1
prev_game_date           1
next_game_date           1
player_first_row         1
player_last_row          1
injury_span_id           1
injury_span_id_orig      1
time_since_last_game     1
non_mlb_days             1
il_days_max              1
il_days_sum              1
prev_non_mlb_days        1
start_date               1
p

**MLB Dataset Observations:**
*   **Description:** The MLB injury database records baseball injuries at a high level of detail, registering stints on the Injured List (IL) and tracking granular anatomical data, precise dates of time missed, and detailed locations of injuries occurring throughout seasons.
*   **Data Source:** https://github.com/ssharpe42/mlb-injury/tree/master/data
*   **Limitations:** High dimensionality (48 columns) could introduce significant complexity and noise making feature selection crucial. The excessive null values across its sparse parameters may require aggressive imputation strategies.
*   **Columns:** It is the most extensive dataset with 48 columns. Key fields include explicit attributes for `injury_type` and `injury_location`, explicit duration measures (`il_days`, `in_season_duration`), temporal context (`year`, `date`), and general text in `notes`.

**Additional Insights:**
*   **Duplicates:** 0 duplicate rows found.
*   **Unique Values:** Shows 11 unique years, aligning with a likely 2010-2020 layout. Contains deep granularity across multiple text fields (e.g. nearly 4,000 unique `notes`), emphasizing the rich but heavy memory profile of this dataset.
* **Injury type column?** Multiple options: `injury_type`, `injury_type_notes`, `injury_location`
* **Games missed/recovery/duration?** Explicit duration tracking in multiple columns: `il_days`, `in_season_duration`, `date_duration`, `duration`
* **How are injuries described?** Extensively broken down into structured details (`injury_type`, `injury_location`, `injury_notes`).
* **Season/year?** Has both `year` and `date`.
* **Cleanest/most chaotic?** Most detailed and feature-rich (48 columns). Might require dropping redundant columns but very comprehensively structured. Size: 9,626 rows, 48 columns.

In [155]:
# Put all four datasets in one dictionary so we can loop through them together.
datasets = {
    "NFL": nfl,
    "NBA": nba,
    "NHL": nhl,
    "MLB": mlb,
}

# Save only the column indexes, because pandas Index supports intersection() and union().
column_indexes = {name: df.columns for name, df in datasets.items()}

# Start from NFL columns, then keep only the column names that also exist in the other datasets.
common_to_all = column_indexes["NFL"]
for columns in column_indexes.values():
    common_to_all = common_to_all.intersection(columns)

# Compare every pair of datasets to find exact column-name matches between them.
pairwise_intersections = {
    f"{left} & {right}": column_indexes[left].intersection(column_indexes[right]).tolist()
    for left, right in combinations(column_indexes, 2)
}

# Build one combined list with every unique column name that appears in any dataset.
all_columns = pd.Index([])
for columns in column_indexes.values():
    all_columns = all_columns.union(columns)

print("Common columns across all four datasets:")
print(common_to_all.tolist())

print("\nExact pairwise column-name intersections:")
for pair, columns in pairwise_intersections.items():
    print(f"{pair}: {columns}")

print(f"\nUnion of all column names ({len(all_columns)} columns):")
print(all_columns.tolist())


Common columns across all four datasets:
[]

Exact pairwise column-name intersections:
NFL & NBA: []
NFL & NHL: ['Player']
NFL & MLB: ['year', 'team']
NBA & NHL: ['Team']
NBA & MLB: []
NHL & MLB: []

Union of all column names (69 columns):
['Acquired', 'Cap Hit', 'Chip', 'Date', 'Games Missed', 'Games Missed.1', 'Injury Type', 'Notes', 'Player', 'Player_id', 'Position', 'Relinquished', 'Season', 'Team', 'Unnamed: 0', '__index_level_0__', 'activated', 'cum_season_days', 'date', 'date_duration', 'dt', 'dtd', 'duration', 'earliest_injury', 'effectiveDate', 'game', 'game_date', 'game_type', 'games_in_season', 'il_days', 'il_days_max', 'il_days_sum', 'in_season_duration', 'injury_date', 'injury_id', 'injury_location', 'injury_location_notes', 'injury_notes', 'injury_span_id', 'injury_span_id_orig', 'injury_type', 'injury_type_notes', 'injury_types', 'latest_injury', 'name', 'next_game_date', 'next_game_season_days', 'next_is_game', 'non_mlb_days', 'notes', 'num_games_injured', 'num_games_mi

### Dataset Comparison
*   **Largest Dataset:** **NBA** leads in total records (27,105 rows), followed by **NHL** (24,498 rows) and **NFL** (17,389 rows). In terms of features (columns), **MLB** is the absolute leader with 48 columns per record (though it only has 9,626 rows).
*   **Cleanest (Structured):** **MLB** and **NFL**. MLB offers the richest and most detailed structure for analysis due to multiple columns for injury duration and location. NFL is also very well-structured and straightforward to work with.
*   **Most Chaotic:** Unquestionably **NBA**. The lack of categorical columns means that serious text parsing and NLP will be required on the `Notes` column to extract the injury types and missed time. NHL also has minor issues (such as the duplicate/unclear `Games Missed.1` column), but it is much better structured than NBA.
*   **Column-name overlap:** Using pandas-style `Index.intersection()` and `Index.union()` logic, there are no exact column names shared by all four datasets because naming conventions differ by case and wording. Exact pairwise overlaps are limited: **NFL** and **MLB** share `year` and `team`; **NBA** and **NHL** share `Team`; and **NFL** and **NHL** share `Player`. The strongest derivative or semantic matches are player identifiers (`Player_id`/`player_id`, `Player`/`name`, `Relinquished`), team fields (`team`, `Team`, `team_id`, `team_name`), season/date fields (`year`, `Season`, `Date`, `date`, `injury_date`, `game_date`, `start_date`), injury descriptions (`injury_types`, `Injury Type`, `injury_type`, `injury_location`, `Notes`, `notes`), and duration fields (`num_games_missing`, `num_games_injured`, `Games Missed`, `Games Missed.1`, `il_days`, `duration`, `in_season_duration`). Different or dataset-specific names include NHL financial columns (`Cap Hit`, `Chip`), MLB event/status columns (`dtd`, `activated`, `transfer`, game-date context fields), NBA transaction columns (`Acquired`, `Relinquished`), and NFL season aggregate fields (`games_in_season`, `earliest_injury`, `latest_injury`). The union of columns is therefore much larger than the intersection, so standardization should map derivative names into common fields before merging.

### Preliminary Conclusion

The datasets vary significantly in structure and level of detail.
The NFL dataset appears more structured, while the NHL dataset contains more inconsistencies.
Standardization will be required before merging the datasets.


The datasets originate from different sources and may use varying definitions and reporting standards for injuries, which should be considered when interpreting the results.